<font color = red size = 5>立此为证</font>

             语文	数学	Python  
        
张三	期中	  50   132	 10  


      期末   117	 22	   9  

李四	期中	  76   113	 50  
 
     期末	    8	23	  19  

活齐  期中	   2  86    70  

      期末    148  50  141  


In [106]:
import numpy as np
import pandas as pd
from pandas import Series, DataFrame

In [107]:
df1 = DataFrame(np.random.randint(0, 150, size = (6, 3)),
               columns = ['Python','C++','Math'],
               index = pd.MultiIndex.from_product([['张三', '李四', 'A王二'],['期中', '期末']]))
df1

Python  C++  Math
张三  期中     113   39    43
    期末     105   84    95
李四  期中     128  139    85
    期末     102  147   129
A王二 期中      68  142    26
    期末      39  125    57

In [5]:
#多层索引时，因为排序问题将会出现问题UnsortedIndexError    
df1.loc['张三':'A王二']

UnsortedIndexError: 'Key length (1) was greater than MultiIndex lexsort depth (0)'

In [6]:
df2 = DataFrame(np.random.randint(0, 150, size = (3, 3)),
               columns = ['Python','C++','Math'],
               index = ['张三', '李四', 'A王二'])
df2

Python  C++  Math
张三      140   82   112
李四       12  146    49
A王二      95   90   120

In [7]:
df2.loc['张三':'A王二']

Python  C++  Math
张三      140   82   112
李四       12  146    49
A王二      95   90   120

In [8]:
df3 = df1.sort_index()
df3

Python  C++  Math
A王二 期中      61  104     4
    期末      96  124    24
张三  期中      36  128   108
    期末      81   89     0
李四  期中      23  102    61
    期末     132   36    63

In [9]:
df3.loc['张三':'A王二']

Empty DataFrame
Columns: [Python, C++, Math]
Index: []

# pandas数据处理

## 1、删除重复元素

使用duplicated()函数检测重复的行，返回元素为布尔类型的Series对象，每个元素对应一行，如果该行不是第一次出现，则元素为True

In [12]:
df = DataFrame({'color':['red', 'bule', 'yellow', 'red'],
               'price':[10, 15, 20, 20]
               })
df

color  price
0     red     10
1    bule     15
2  yellow     20
3     red     20

使用drop_duplicates()函数删除重复的行

In [11]:
df.drop_duplicates()

color  price
0     red     10
1    bule     15
2  yellow     20

如果使用pd.concat([df1,df2],axis = 1)生成新的DataFrame，新的df中columns相同，使用duplicate()和drop_duplicates()都会出问题

## 2. 映射

映射的含义：创建一个映射关系列表，把values元素和一个特定的标签或者字符串绑定

需要使用字典：

`map = {
    'label1':'value1',
    'label2':'value2',
    ...
    }
`

包含三种操作：

- replace()函数：替换元素
- 最重要：map()函数：新建一列
- rename()函数：替换索引

In [13]:
df

color  price
0     red     10
1    bule     15
2  yellow     20
3     red     20

### 1) replace()函数：替换元素

使用replace()函数，对values进行替换操作

首先定义一个字典

调用.replace()

replace还经常用来替换NaN元素

In [15]:
df.replace({10:18,15:50})

color  price
0     red     18
1    bule     50
2  yellow     20
3     red     20

In [18]:
df.replace({10:18,15:30},inplace = True)
df

color  price
0     red     18
1    bule     30
2  yellow     20
3     red     20

============================================

练习19：

    假设张三李四的课表里有满分的情况，老师认为是作弊，把所有满分的情况（包括150,300分）都记0分，如何实现？

============================================

### 2) map()函数：新建一列

使用map()函数，由已有的列生成一个新列

适合处理某一单独的列。

In [20]:
df = DataFrame(np.random.randint(0, 150, size = (100, 3)),columns = ['Python', 'Math', 'En'])
df

Python  Math   En
0       14    77  140
1       27    82   37
2       49    67   45
3      129     2    0
4       55   133   95
..     ...   ...  ...
95      44   112   53
96     125   111   70
97     115   108  102
98     125    50  124
99     103    37   52

[100 rows x 3 columns]

In [24]:
df['Python'] = df['Python'].map(lambda x: x + 10)
df

Python  Math   En
0       54    77  140
1       67    82   37
2       89    67   45
3      169     2    0
4       95   133   95
..     ...   ...  ...
95      84   112   53
96     165   111   70
97     155   108  102
98     165    50  124
99     143    37   52

[100 rows x 3 columns]

In [34]:
def convert_level(x):
    if x < 60:
        return '不及格'
    elif x < 90:
        return '及格'
    else:
        return '优秀'
    
df['level'] = df.mean(axis = 1).map(convert_level)
df

Python  Math   En level
0       54    77  140    优秀
1       67    82   37    及格
2       89    67   45    及格
3      169     2    0   不及格
4       95   133   95    优秀
..     ...   ...  ...   ...
95      84   112   53    及格
96     165   111   70    优秀
97     155   108  102    优秀
98     165    50  124    优秀
99     143    37   52    及格

[100 rows x 4 columns]

In [36]:
df['Java'] = df['Python'].map(lambda x: x+10)
df


Python  Math   En level  Java
0       54    77  140    优秀    64
1       67    82   37    及格    77
2       89    67   45    及格    99
3      169     2    0   不及格   179
4       95   133   95    优秀   105
..     ...   ...  ...   ...   ...
95      84   112   53    及格    94
96     165   111   70    优秀   175
97     155   108  102    优秀   165
98     165    50  124    优秀   175
99     143    37   52    及格   153

[100 rows x 5 columns]

In [38]:
df['Chinese'] = df['Python'].map(lambda x: x+20)
df

Python  Math   En level  Java  Chinese
0       54    77  140    优秀    64       74
1       67    82   37    及格    77       87
2       89    67   45    及格    99      109
3      169     2    0   不及格   179      189
4       95   133   95    优秀   105      115
..     ...   ...  ...   ...   ...      ...
95      84   112   53    及格    94      104
96     165   111   70    优秀   175      185
97     155   108  102    优秀   165      175
98     165    50  124    优秀   175      185
99     143    37   52    及格   153      163

[100 rows x 6 columns]

In [39]:
df.corr()

Python      Math        En      Java   Chinese
Python   1.000000 -0.150899 -0.012577  1.000000  1.000000
Math    -0.150899  1.000000  0.070003 -0.150899 -0.150899
En      -0.012577  0.070003  1.000000 -0.012577 -0.012577
Java     1.000000 -0.150899 -0.012577  1.000000  1.000000
Chinese  1.000000 -0.150899 -0.012577  1.000000  1.000000

In [ ]:
df['Python'].map(lambda x: x+10)

仍然是新建一个字典

map()函数中可以使用lambda函数

transform()和map()类似

使用map()函数新建一个新列

In [40]:
df.head()

Python  Math   En level  Java  Chinese
0      54    77  140    优秀    64       74
1      67    82   37    及格    77       87
2      89    67   45    及格    99      109
3     169     2    0   不及格   179      189
4      95   133   95    优秀   105      115

In [42]:
df['Chem'] = df['Python'].map(lambda x: x / 2) + df['Java'].map (lambda x: x/2)
df

Python  Math   En level  Java  Chinese   Chem
0       54    77  140    优秀    64       74   59.0
1       67    82   37    及格    77       87   72.0
2       89    67   45    及格    99      109   94.0
3      169     2    0   不及格   179      189  174.0
4       95   133   95    优秀   105      115  100.0
..     ...   ...  ...   ...   ...      ...    ...
95      84   112   53    及格    94      104   89.0
96     165   111   70    优秀   175      185  170.0
97     155   108  102    优秀   165      175  160.0
98     165    50  124    优秀   175      185  170.0
99     143    37   52    及格   153      163  148.0

[100 rows x 7 columns]

In [44]:
df['Python'].transform(lambda x: x-10)

0      44
1      57
2      79
3     159
4      85
     ... 
95     74
96    155
97    145
98    155
99    133
Name: Python, Length: 100, dtype: int64

In [46]:
df['price'] = np.random.randint(150,898,size = 100)

In [47]:
df['count'] = np.random.randint(2, 10, size = 100)

In [48]:
df.head(b)

Python  Math   En level  Java  Chinese   Chem  price  count
0      54    77  140    优秀    64       74   59.0    587      5
1      67    82   37    及格    77       87   72.0    874      9
2      89    67   45    及格    99      109   94.0    397      3
3     169     2    0   不及格   179      189  174.0    633      9
4      95   133   95    优秀   105      115  100.0    484      4

In [52]:
df.count

<bound method DataFrame.count of     Python  Math   En level  Java  Chinese   Chem  price  count
0       54    77  140    优秀    64       74   59.0    587      5
1       67    82   37    及格    77       87   72.0    874      9
2       89    67   45    及格    99      109   94.0    397      3
3      169     2    0   不及格   179      189  174.0    633      9
4       95   133   95    优秀   105      115  100.0    484      4
..     ...   ...  ...   ...   ...      ...    ...    ...    ...
95      84   112   53    及格    94      104   89.0    297      4
96     165   111   70    优秀   175      185  170.0    505      7
97     155   108  102    优秀   165      175  160.0    377      7
98     165    50  124    优秀   175      185  170.0    560      9
99     143    37   52    及格   153      163  148.0    438      3

[100 rows x 9 columns]>

In [54]:
df['total'] = df.price * df['count']

In [55]:
df.head()

Python  Math   En level  Java  Chinese   Chem  price  count  total
0      54    77  140    优秀    64       74   59.0    587      5   2935
1      67    82   37    及格    77       87   72.0    874      9   7866
2      89    67   45    及格    99      109   94.0    397      3   1191
3     169     2    0   不及格   179      189  174.0    633      9   5697
4      95   133   95    优秀   105      115  100.0    484      4   1936

============================================

练习20：

    新增两列，分别为张三、李四的成绩状态，如果分数低于90，则为"failed"，如果分数高于120，则为"excellent"，其他则为"pass"
    
    【提示】使用函数作为map的参数

============================================

### 3) rename()函数：替换索引

仍然是新建一个字典

使用rename()函数替换行索引

In [56]:
df.rename({0:'A'})

Python  Math   En level  Java  Chinese   Chem  price  count  total
A       54    77  140    优秀    64       74   59.0    587      5   2935
1       67    82   37    及格    77       87   72.0    874      9   7866
2       89    67   45    及格    99      109   94.0    397      3   1191
3      169     2    0   不及格   179      189  174.0    633      9   5697
4       95   133   95    优秀   105      115  100.0    484      4   1936
..     ...   ...  ...   ...   ...      ...    ...    ...    ...    ...
95      84   112   53    及格    94      104   89.0    297      4   1188
96     165   111   70    优秀   175      185  170.0    505      7   3535
97     155   108  102    优秀   165      175  160.0    377      7   2639
98     165    50  124    优秀   175      185  170.0    560      9   5040
99     143    37   52    及格   153      163  148.0    438      3   1314

[100 rows x 10 columns]

In [58]:
df.rename({'En':'英语'}, inplace = True
         )

In [59]:
df

Python  Math   En level  Java  Chinese   Chem  price  count  total
0       54    77  140    优秀    64       74   59.0    587      5   2935
1       67    82   37    及格    77       87   72.0    874      9   7866
2       89    67   45    及格    99      109   94.0    397      3   1191
3      169     2    0   不及格   179      189  174.0    633      9   5697
4       95   133   95    优秀   105      115  100.0    484      4   1936
..     ...   ...  ...   ...   ...      ...    ...    ...    ...    ...
95      84   112   53    及格    94      104   89.0    297      4   1188
96     165   111   70    优秀   175      185  170.0    505      7   3535
97     155   108  102    优秀   165      175  160.0    377      7   2639
98     165    50  124    优秀   175      185  170.0    560      9   5040
99     143    37   52    及格   153      163  148.0    438      3   1314

[100 rows x 10 columns]

## 3. 异常值检测和过滤

使用describe()函数查看每一列的描述性统计量

使用std()函数可以求得DataFrame对象每一列的标准差

In [61]:
#正态分布
df = DataFrame(np.random.randn(100000,3),columns = ['red', 'green','blue'])
df

red     green      blue
0      0.680888 -0.298569 -0.386061
1      1.428152  1.126917 -0.195375
2      0.768562  2.295910  0.158464
3      1.819718 -1.614846  0.863118
4     -0.179418  0.019371 -0.073688
...         ...       ...       ...
99995  0.801145  0.891347  0.633236
99996 -0.360687 -0.540382 -0.221256
99997 -0.731153 -0.523084  0.069426
99998  0.146211 -1.245166  0.028193
99999  0.721185  0.033521  0.777517

[100000 rows x 3 columns]

In [63]:
df.abs() - df.mean() > 3 * df.std()

red  green   blue
0      False  False  False
1      False  False  False
2      False  False  False
3      False  False  False
4      False  False  False
...      ...    ...    ...
99995  False  False  False
99996  False  False  False
99997  False  False  False
99998  False  False  False
99999  False  False  False

[100000 rows x 3 columns]

In [70]:
cond = df.abs() - df.mean() > 3 * df.std()
cond.sum()

red      244
green    288
blue     284
dtype: int64

In [71]:
cond = cond.any(axis = 1)
df[cond]

red     green      blue
25    -0.995501  3.035962 -0.022944
69     0.109961 -3.051660  0.930003
95    -0.140262 -0.149699  3.013783
109    0.175239 -3.636037 -1.361508
602    3.082179  1.097644  0.912568
...         ...       ...       ...
99671 -3.342718  1.369030  0.538269
99680 -3.371526 -0.459973  0.698370
99694  0.297153  3.660528  1.902792
99741 -1.622467  0.255037 -2.995002
99947  0.634683  3.075461 -0.496950

[811 rows x 3 columns]

In [75]:
df2 = df.drop(labels=df[cond].index)

In [74]:
df.shape

(100000, 3)

In [76]:
df2.shape

(99189, 3)

根据每一列的标准差，对DataFrame元素进行过滤。

借助any()函数, 测试是否有True，有一个或以上返回True，反之返回False

对每一列应用筛选条件,去除标准差太大的数据

删除特定索引df.drop(labels,inplace = True)

============================================

练习21：

    新建一个形状为10000*3的标准正态分布的DataFrame(np.random.randn)，去除掉所有满足以下情况的行：其中任一元素绝对值大于3倍标准差

============================================

## 4. 排序

使用.take()函数排序

可以借助np.random.permutation()函数随机排序

In [77]:
df.sort_values(by = 'red', ascending = False)

red     green      blue
60298  3.934460 -0.648550  0.612789
78296  3.919186  0.735172  1.842114
34562  3.894454  0.628953  0.013888
1385   3.863666 -1.124287 -0.677137
52367  3.822057 -0.894033  0.478079
...         ...       ...       ...
56707 -3.873804  1.271019 -1.813218
9152  -3.907315 -1.545100 -0.268103
63650 -4.042376  1.003482 -1.032063
5512  -4.065315  0.262826  1.369788
16707 -4.899485 -0.741991  0.335100

[100000 rows x 3 columns]

### 随机抽样

当DataFrame规模足够大时，直接使用np.random.randint()函数，就配合take()函数实现随机抽样

In [80]:
indexs = np.random.randint(0, 10000, size = 1000)
indexs

array([1365, 4749, 5467, 2597, 7386, 3051, 9576, 8221, 9723, 2903, 6782,
       8572, 6212, 1922, 4513, 7405, 4198, 3103, 2192,  242, 9274, 5781,
       5223,   15, 5565, 3487, 2052, 8866, 6476, 2756, 7599, 2108, 5296,
       8683, 8009, 6836, 1328, 6648, 6174, 8704, 4764, 6308, 5818, 5960,
       1672, 9759,   35, 7464, 4439, 8009, 3367, 8675, 4330, 7484, 1702,
       7179, 1817, 2949,  714, 6213, 1109, 3764, 7020, 9507, 1830, 8126,
       9637, 2428, 2607,  330,  309, 4371, 4774, 7198, 6832, 4613,  361,
       4070, 2989, 6614, 5249, 7640, 4608, 6309, 7433, 2963, 1827, 9820,
        923, 2816, 6071, 1612, 4436, 8193, 7220, 4989, 9791, 5795, 2280,
       5021, 7804, 7719, 7472, 9032, 1133, 4696, 8446, 7951, 3729, 6009,
        212, 2471, 5959, 5475, 9229, 6432, 5987, 9529, 8395, 8814, 7544,
       8254,  341, 7487, 9973, 3175, 8301, 2613, 5059, 9243,  760, 6351,
       2784,  438, 7193, 3559, 9205, 3250, 5057, 7270, 4860, 7444, 1734,
       1258, 9915, 2134, 4755,  489, 7974, 1372, 75

In [84]:
df.take(indexs)

red     green      blue
1365 -0.674749  0.076908 -1.414229
4749  0.571783  0.336247  1.051226
5467  0.611791 -0.570785  0.374208
2597 -0.841600  1.135647  1.512508
7386  0.519257  0.699713 -0.774952
...        ...       ...       ...
6520  2.112557  0.541442 -0.174947
3457  0.619307 -0.987725  0.795481
5231 -0.085358 -0.372113  1.298449
5345  0.834267 -0.486620  2.273029
288   1.226907  0.957822 -2.154550

[1000 rows x 3 columns]

In [85]:
df.shape

(100000, 3)

============================================
练习22：

   假设有张三李四王老五的期中考试成绩ddd2，对着三名同学随机排序

============================================

## 5. 数据聚合【重点】

数据聚合是数据处理的最后一步，通常是要使每一个数组生成一个单一的数值。

数据分类处理：

 - 分组：先把数据分为几组
 - 用函数处理：为不同组的数据应用不同的函数以转换数据
 - 合并：把不同组得到的结果合并起来
 
数据分类处理的核心：
     groupby()函数

In [90]:
df = DataFrame({'item': np.random.randint(0,10,size = 100),
               'saler': np.random.randint(0, 10, size = 100),
               'weight':np.random.randint(30, 300, size  = 100),
                'price':np.random.randint(1, 20, size = 100)
               })
df

item  saler  weight  price
0      3      9     186      2
1      1      4     262     13
2      6      5     144     14
3      2      0     289      9
4      4      3      81     18
..   ...    ...     ...    ...
95     5      7     255     15
96     4      6     141     10
97     1      6     141     18
98     1      6     168     15
99     2      2      80     19

[100 rows x 4 columns]

In [91]:
def convert_item(x):
    if x< 2:
        return '萝卜'
    elif x<7:
        return '白菜'
    else:
        return '青椒'
df['item'] = df['item'].map(convert_item)

In [93]:
def convert_saler(x):
    if x<5:
        return '张传龙'
    elif x<7:
        return '孙悟空'
    else:
        return '唐僧'
df['saler'] = df['saler'].map(convert_saler)

In [94]:
def convert_weight(x):
    if x< 80:
        return 50
    elif x <150:
        return 130
    else :
        return 150
df['weight'] = df['weight'].transform(convert_weight)

In [95]:
df

item saler  weight  price
0    白菜    唐僧     150      2
1    萝卜   张传龙     150     13
2    白菜   孙悟空     130     14
3    白菜   张传龙     150      9
4    白菜   张传龙     130     18
..  ...   ...     ...    ...
95   白菜    唐僧     150     15
96   白菜   孙悟空     130     10
97   萝卜   孙悟空     130     18
98   萝卜   孙悟空     150     15
99   白菜   张传龙     130     19

[100 rows x 4 columns]

In [102]:
#统计一月的销售重量
all_sale = df.groupby(['saler'])['weight']
all_sale

In [133]:
#分组之后需要计算
all_sale.sum()

saler
唐僧     2810
孙悟空    2490
张传龙    6560
Name: weight, dtype: int64

In [99]:
all_sale.min()

saler
唐僧     50
孙悟空    50
张传龙    50
Name: weight, dtype: int64

In [101]:
all_sale.apply(np.sum)

saler
唐僧     2810
孙悟空    2490
张传龙    6560
Name: weight, dtype: int64

In [110]:
all_sale.count()

saler
唐僧     23
孙悟空    21
张传龙    56
Name: weight, dtype: int64

In [114]:
all_sale

In [116]:
def count(x):
    return x.size

In [117]:
all_sale.apply(count)

saler
唐僧     23
孙悟空    21
张传龙    56
Name: weight, dtype: int64

In [119]:
def analysis(x):
    return (x.size, x.sum(), x.mean(), x.min)
all_sale.apply(analysis)


saler
唐僧     (23, 2810, 122.17391304347827, <bound method S...
孙悟空    (21, 2490, 118.57142857142857, <bound method S...
张传龙    (56, 6560, 117.14285714285714, <bound method S...
Name: weight, dtype: object

In [123]:
df

item saler  weight  price
0    白菜    唐僧     150      2
1    萝卜   张传龙     150     13
2    白菜   孙悟空     130     14
3    白菜   张传龙     150      9
4    白菜   张传龙     130     18
..  ...   ...     ...    ...
95   白菜    唐僧     150     15
96   白菜   孙悟空     130     10
97   萝卜   孙悟空     130     18
98   萝卜   孙悟空     150     15
99   白菜   张传龙     130     19

[100 rows x 4 columns]

In [124]:
g = df.groupby(['saler','item'])
g

In [127]:
np.round(g.mean())


weight  price
saler item               
唐僧    白菜     130.0    8.0
      萝卜     112.0    9.0
      青椒     123.0   12.0
孙悟空   白菜     128.0   11.0
      萝卜     120.0   12.0
      青椒      95.0    6.0
张传龙   白菜     112.0   12.0
      萝卜      98.0   10.0
      青椒     134.0   10.0

In [128]:
g.min()

weight  price
saler item               
唐僧    白菜        50      1
      萝卜        50      2
      青椒        50      2
孙悟空   白菜        50      4
      萝卜        50      3
      青椒        50      1
张传龙   白菜        50      1
      萝卜        50      1
      青椒        50      3

In [131]:
g

In [132]:
#价格求平均，销量求和
g.agg(['sum', 'mean'])

weight             price           
              sum        mean   sum       mean
saler item                                    
唐僧    白菜     1170  130.000000    70   7.777778
      萝卜      900  112.500000    69   8.625000
      青椒      740  123.333333    73  12.166667
孙悟空   白菜     1150  127.777778   100  11.111111
      萝卜      960  120.000000    97  12.125000
      青椒      380   95.000000    22   5.500000
张传龙   白菜     3360  112.000000   355  11.833333
      萝卜      780   97.500000    77   9.625000
      青椒     2420  134.444444   178   9.888889

使用.groups属性查看各行的分组情况：

============================================

练习23：

   假设菜市场张大妈在卖菜，有以下属性：
   
   菜品(item)：萝卜，白菜，辣椒，冬瓜
   
   颜色(color)：白，青，红
   
   重量(weight)
   
   价格(price)
   
1. 要求以属性作为列索引，新建一个ddd
2. 对ddd进行聚合操作，求出颜色为白色的价格总和
3. 对ddd进行聚合操作，求出萝卜的所有重量(包括白萝卜，胡萝卜，青萝卜）以及平均价格
4. 使用merge合并总重量及平均价格

============================================

## 6.0 高级数据聚合

可以使用pd.merge()函数将聚合操作的计算结果添加到df的每一行  
使用groupby分组后调用加和等函数进行运算，让后最后可以调用add_prefix()，来修改列名

### 可以使用transform和apply实现相同功能

在transform或者apply中传入函数即可

### transform()与apply()函数还能传入一个函数或者lambda

`
df = DataFrame({'color':['white','black','white','white','black','black'],
               'status':['up','up','down','down','down','up'],
               'value1':[12.33,14.55,22.34,27.84,23.40,18.33],
               'value2':[11.23,31.80,29.99,31.18,18.25,22.44]})
`

apply的操作对象，也就是传给lambda的参数是整列的数组

============================================

练习24：

   使用transform与apply实现练习23的功能

============================================